File: ra-idp-go/internal/oauth2/ports/session_store.go
package ports

import (
	"context"

	"ra-idp-go/internal/spec"
)

type SessionStore interface {
	Save(ctx context.Context, s *spec.LoginSession) error
	Find(ctx context.Context, sessionID string) (*spec.LoginSession, error)
	Delete(ctx context.Context, sessionID string) error
}

File: ra-idp/adapters/persistence/memory/session-store.ts
/**
 * Layer 4 — Adapter Layer (in-memory SessionStore)
 */

import type { LoginSession, SessionStore } from '../../../src/authentication/ports/session-store'

export class InMemorySessionStore implements SessionStore {
  private readonly store = new Map<string, LoginSession>()

  async find(id: string): Promise<LoginSession | null> {
    const v = this.store.get(id)
    if (!v) return null
    if (Date.parse(v.expires_at) <= Date.now()) {
      this.store.delete(id)
      return null
    }
    return { ...v }
  }

  async save(session: LoginSession): Promise<void> {
    this.store.set(session.id, { ...session })
  }

  async delete(id: string): Promise<void> {
    this.store.delete(id)
  }
}
